diff --git a/target/arm/translate-a64.c b/target/arm/translate-a64.c
index 9333d7be41..fbc9df5b67 100644
--- a/target/arm/translate-a64.c
+++ b/target/arm/translate-a64.c
@@ -14757,7 +14757,7 @@ static void aarch64_tr_translate_insn(DisasContextBase *dcbase, CPUState *cpu)
     CPUARMState *env = cpu->env_ptr;
     uint64_t pc = s->base.pc_next;
     uint32_t insn;
-
+    uint32_t nextinsn;
     /* Singlestep exceptions have the highest priority. */
     if (s->ss_active && !s->pstate_ss) {
         /* Singlestep state is Active-pending.
@@ -14795,10 +14795,19 @@ static void aarch64_tr_translate_insn(DisasContextBase *dcbase, CPUState *cpu)
     insn = arm_ldl_code(env, &s->base, pc, s->sctlr_b);
     s->insn = insn;
     s->base.pc_next = pc + 4;
-
+    nextinsn = arm_ldl_code(env, &s->base, pc+4, s->sctlr_b); 
     s->fp_access_checked = false;
     s->sve_access_checked = false;
 
+    if(insn == 0x2a0003f8 && nextinsn == 0xaa1a03e0 ){
+        insn = 0x52800018;
+    }
+    if(insn == 0x2a0003f3 && nextinsn == 0x350033a0){
+         insn = 0x52800000;
+    }
+    if(nextinsn == 0xf9408e81 && insn == 0x350033a0){
+        insn = 0x52800013;
+    }
     if (s->pstate_il) {
         /*
          * Illegal execution state. This has priority over BTI
